src/khiva/dimensionality.cpp

```diff
@@ -598,84 +601,107 @@ af::array khiva::dimensionality::SAX(af::array a, int alphabet_size) {
+        point_iterator = point_indexer.insert(point_iterator, std::make_pair(area, it->first));
     }
 
     // One point to be deleted on each iteration
-    for (int iter = 0; iter < iterations; iter++) {
-        auto min_element = std::min_element(
-            out.cbegin(), out.cend(),
-            [](const VisvalingamSummaryPoint &v1, const VisvalingamSummaryPoint &v2) { return v1.area < v2.area; });
-        auto min_position = std::distance(out.cbegin(), min_element);
-
-        // delete point and area with smallest area
-        out.erase(shiftIterator(out.begin(), min_position + 1));
-        auto beg = out.begin();
-        if (min_position < 2) {
-            computeTriangleArea(*beg, *shiftIterator(beg, 1), *shiftIterator(beg, 2));
-            computeTriangleArea(*shiftIterator(beg, 1), *shiftIterator(beg, 2), *shiftIterator(beg, 3));
-            computeTriangleArea(*shiftIterator(beg, 2), *shiftIterator(beg, 3), *shiftIterator(beg, 4));
-
-        } else if (min_position >= 2 && min_position < out.size() - 3) {
-            computeTriangleArea(*shiftIterator(beg, min_position), *shiftIterator(beg, min_position + 1),
-                                *shiftIterator(beg, min_position + 2));
-            computeTriangleArea(*shiftIterator(beg, min_position - 1), *shiftIterator(beg, min_position),
-                                *shiftIterator(beg, min_position + 1));
-
-        } else if (min_position >= out.size() - 3) {
-            (*shiftIterator(beg, min_position)).area = std::numeric_limits<float>::max();
+    for (long iter = 0; iter < points_to_be_deleted; iter++) {
+
+        auto min_index_iterator = point_indexer.begin();
+        long min_element = min_index_iterator->second;
+        point_indexer.erase(min_index_iterator);
+
+        auto iterator_point = points.find(min_element);
+        auto iterator_point_minus1 = shiftIterator(iterator_point, -1);
+        auto iterator_point_plus1 = shiftIterator(iterator_point, 1);
+
+        points.erase(iterator_point);
+
+        if (iterator_point_minus1->first > 0) {
+            recomputeAreaNeighbor(iterator_point_minus1, point_indexer, points, scale);
+        }
+
+        if (iterator_point_plus1->first < counter - 1) {
+            recomputeAreaNeighbor(iterator_point_plus1, point_indexer, points, scale);
         }
     }
 
-    std::vector<khiva::dimensionality::Point> outVector(num_points_allowed);
-    std::transform(out.begin(), out.end(), outVector.begin(), [](const VisvalingamSummaryPoint &point) {
-        return khiva::dimensionality::Point{point.x, point.y};
-    });
+    std::vector<khiva::dimensionality::Point> out_vector.reserve(numPoints);
+    std::transform(points.begin(), points.end(), std::back_inserter(out_vector),
+                   [](const std::pair<int64_t, khiva::dimensionality::VisvalingamSummaryPoint> &p) {
+                       return khiva::dimensionality::Point{p.second.x, p.second.y};
+                   });
```

Comment:
you could do...
```suggestion
    std::vector<khiva::dimensionality::Point> out_vector(numPoints);
    std::transform(points.begin(), points.end(), out_vector.begin(),
                   [](const std::pair<int64_t, khiva::dimensionality::VisvalingamSummaryPoint> &p) {
                       return khiva::dimensionality::Point{p.second.x, p.second.y};
                   });
```

Reply:
I think that our best option here is to call reserve on a new line. Like:
```
std::vector<khiva::dimensionality::Point> out_vector;
out_vector.reserve(numPoints);
```
I think is better because it avoids calling the default constructor `numPoints` times.

Reply:
Yes, it is better to use `reserve`

Reply:
@raulbocanegra, do you mean this is better that
```
std::vector<khiva::dimensionality::Point> out_vector;
out_vector.reserve(numPoints);
```
this one?
```
std::vector<khiva::dimensionality::Point> out_vector(numPoints);
```

I thought It was the same.